src/github.com/getlantern/eventual/eventual.go

```diff
@@ -51,34 +56,59 @@ func DefaultGetter(val interface{}) Getter {
 }
 
 func (v *value) Set(val interface{}) {
-	v.updates <- val
+	// Prevent sending on closed channel
+	if atomic.LoadInt32(&v.stopped) == intFalse {
+		v.muUpdates.RLock()
+		v.updates <- val
+		v.muUpdates.RUnlock()
+	}
 }
 
 func (v *value) processUpdates() {
 	for val := range v.updates {
 		v.val.Store(val)
-		if v.gotFirst == FALSE {
+		if v.gotFirst == intFalse {
 			// Signal to blocking callers that we have the first value
 			v.wg.Done()
-			atomic.StoreInt32(&v.gotFirst, TRUE)
+			atomic.StoreInt32(&v.gotFirst, intTrue)
 		}
 	}
+	// Ensure Get() to return when Stop() is called
+	if atomic.LoadInt32(&v.gotFirst) == intFalse {
+		v.wg.Done()
+	}
+}
+
+func (v *value) Stop() {
+	// Prevent closing multiple times
+	if atomic.CompareAndSwapInt32(&v.stopped, intFalse, intTrue) {
+		v.muUpdates.Lock()
+		close(v.updates)
+		v.muUpdates.Unlock()
+	}
 }
 
-func (v *value) Get(timeout time.Duration) (interface{}, bool) {
-	if atomic.LoadInt32(&v.gotFirst) == TRUE {
+// Get waits the value to be set and returns it, or returns nil if times out or
+// Stop() is called. valid will be false in latter case.
+// TODO: Get should happen after Set if no timeout provided.
+func (v *value) Get(timeout time.Duration) (ret interface{}, valid bool) {
+	if atomic.LoadInt32(&v.gotFirst) == intTrue {
 		// Short-cut used once value has been set, to avoid extra goroutine
 		return v.val.Load(), true
 	}
 
-	valCh := make(chan interface{})
+	// Make it buffered so if the caller no longer receives on the channel, we
+	// can still exit the goroutine.
+	valCh := make(chan interface{}, 1)
```

Comment:
@fffw You said that there's a problem with the below goroutine leaking, but this looks like you fixed it.  Are there still situations where the goroutine can leak?